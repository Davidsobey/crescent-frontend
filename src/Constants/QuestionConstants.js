const QuestionConstants = {
  CREATE_REQUEST: 'CREATE_MODULE_REQUEST',
  CREATE_SUCCESS: 'CREATE_QUESTION_SUCCESS',
  CREATE_FAILURE: 'CREATE_QUESTION_FAILURE',

  GETALL_REQUEST: 'GETALL_QUESTION_REQUEST',
  GETALL_SUCCESS: 'GETALL_QUESTION_SUCCESS',
  GETALL_FAILURE: 'GETALL_QUESTION_FAILURE',

  GETBYID_REQUEST: 'GETBYID_QUESTION_REQUEST',
  GETBYID_SUCCESS: 'GETBYID_QUESTION_SUCCESS',
  GETBYID_FAILURE: 'GETBYID_QUESTION_FAILURE',

  LOAD_QUESTIONS_REQUEST: 'LOAD_QUESTIONS_REQUEST',
  LOAD_QUESTIONS_SUCCESS: 'LOAD_QUESTIONS_SUCCESS',
  LOAD_QUESTIONS_FAILURE: 'LOAD_QUESTIONS_FAILURE',
};

export default QuestionConstants;
