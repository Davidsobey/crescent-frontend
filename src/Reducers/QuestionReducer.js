import { isNumber } from 'util';

import QuestionConstants from '../Constants/QuestionConstants';
import UserConstants from '../Constants/UserConstants';
import OptionConstants from '../Constants/OptionConstants';

function filterById(id, delId) {
  if (isNumber(id) && id !== 0 && id !== delId) {
    return true;
  }
  return false;
}

function findIndex(state, obj) {
  let id = 0;
  let position = 0;
  state.map((object) => {
    if (object.id === obj.id) {
      /* eslint prefer-destructuring: ["error", {VariableDeclarator: {object: true}}] */
      id = position;
    }
    position += 1;
    return position;
  });
  return id;
}

function QuestionReducer(state = {}, action) {
  switch (action.type) {
    case QuestionConstants.CREATE_REQUEST:
      return Object.assign({}, state, {
        question: action.QuestionName,
        creating: true,
      });
    case QuestionConstants.CREATE_SUCCESS:
      return Object.assign({}, state, {
        question: action.questionResponse,
        creating: false,
      });
    case QuestionConstants.CREATE_FAILURE:
      return Object.assign({}, state, {
        question: {},
        creating: false,
      });
    case QuestionConstants.GETALL_REQUEST:
      return Object.assign({}, state, {
        questions: [],
        loading: true,
      });
    case QuestionConstants.GETALL_SUCCESS:
      return Object.assign({}, state, {
        questions: action.questions,
        loading: false,
      });
    case QuestionConstants.GETALL_FAILURE:
      return Object.assign({}, state, {
        questions: [],
        loading: false,
      });
    case QuestionConstants.LOADQUESTIONS_REQUEST:
      return Object.assign({}, state, {
        questions: [],
        loading: true,
      });
    case QuestionConstants.LOADQUESTIONS_SUCCESS:
      return Object.assign({}, state, {
        questions: action.questions,
        loading: false,
      });
    case QuestionConstants.LOADQUESTIONS_FAILURE:
      return Object.assign({}, state, {
        questions: [],
        loading: false,
      });
    case QuestionConstants.GETBYID_REQUEST:
      return Object.assign({}, state, {
        loading: true,
      });
    case QuestionConstants.GETBYID_SUCCESS:
      return Object.assign({}, state, {
        question: action.question,
      });
    case QuestionConstants.GETBYID_FAILURE:
      return Object.assign({}, state, {
        question: {},
      });
    case QuestionConstants.DELETE_REQUEST:
      return Object.assign({}, state, {
        loading: true,
      });
    case QuestionConstants.DELETE_SUCCESS:
      return Object.assign({}, state, {
        questions: state.questions.filter(obj => filterById(obj.id, action.id)),
        loading: false,
      });
    case QuestionConstants.DELETE_FAILURE:
      return state;
    case QuestionConstants.OPTIONS_REQUEST:
      return Object.assign({}, state, {
        question: action.obj,
        loading: true,
      });
    case QuestionConstants.OPTIONS_SUCCESS:
      return Object.assign({}, state, {
        options: action.questionOptions,
      });
    case QuestionConstants.OPTIONS_FAILURE:
      return Object.assign({}, state, {
        question: {},
        options: {},
      });
    case QuestionConstants.OPTIONS_ISANSWER_REQUEST:
      return Object.assign({}, state, {
        loading: true,
      });
    case QuestionConstants.OPTIONS_ISANSWER_SUCCESS:
      return Object.assign({}, state, {
        options: {
          ...Object.values(state.options),
          [findIndex(Object.values(state.options), action.values)]: {
            ...action.values,
            isAnswer: !action.values.isAnswer,
          },
        },
      });
    case QuestionConstants.OPTIONS_ISANSWER_FAILURE:
      return state;
    case QuestionConstants.DELETE_QO_REQUEST:
      return Object.assign({}, state, {
        loading: true,
      });
    case QuestionConstants.DELETE_QO_SUCCESS:
      return Object.assign({}, state, {
        options: state.options.filter(obj => filterById(obj.id, action.id)),
        loading: false,
      });
    case QuestionConstants.DELETE_QO_FAILURE:
      return state;
    case OptionConstants.CREATE_REQUEST:
      return state;
    case OptionConstants.CREATE_SUCCESS:
      return Object.assign({}, state, {
        options: state.options
          ? state.options.concat(action.option)
          : [action.option],
      });
    case OptionConstants.CREATE_FAILURE:
      return state;
    case UserConstants.LOGOUT:
      return {
        question: {},
        questions: [],
      };
    case QuestionConstants.CLEAR_QUESTION:
      return {
        question: {},
        questions: [],
      };
    default:
      return state;
  }
}

export default QuestionReducer;
