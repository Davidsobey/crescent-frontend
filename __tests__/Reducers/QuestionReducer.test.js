// Reducer to be tested
import QuestionReducer from '../../src/Reducers/QuestionReducer';
import QuestionConstants from '../../src/Constants/QuestionConstants';

describe('QuestionReducer', () => {
  test('CREATE_SUCCESS', () => {
    const action = {
      type: QuestionConstants.CREATE_SUCCESS,
      questionResponse: {},
    };

    expect(QuestionReducer(undefined, action)).toMatchSnapshot();
  });

  test('GETALL_SUCCESS', () => {
    const action = {
      type: QuestionConstants.GETALL_SUCCESS,
      questions: [],
    };

    expect(QuestionReducer(undefined, action)).toMatchSnapshot();
  });

  test('GETBYID_SUCCESS', () => {
    const action = {
      type: QuestionConstants.GETBYID_SUCCESS,
      question: {id:1, name:'111'},
    };

    expect(QuestionReducer(undefined, action)).toMatchSnapshot();
  });

  test('LOADQUESTIONS_SUCCESS', () => {
    const action = {
      type: QuestionConstants.LOADQUESTIONS_SUCCESS,
      questions: [],
    };

    expect(QuestionReducer(undefined, action)).toMatchSnapshot();
  });

  test('DELETE_SUCCESS', () => {
    const action = {
      type: QuestionConstants.DELETE_SUCCESS,
      id: 1,
    };

    const state = {
      questions: []
    };

    expect(QuestionReducer(state, action)).toMatchSnapshot();
  });

  test('OPTIONS_SUCCESS', () => {
    const action = {
      type: QuestionConstants.OPTIONS_SUCCESS,
      questionOptions: [],
    };

    expect(QuestionReducer(undefined, action)).toMatchSnapshot();
  });

  test('OPTIONS_ISANSWER_SUCCESS', () => {
    const action = {
      type: QuestionConstants.OPTIONS_ISANSWER_SUCCESS,
      values: {isAnswer: false},
    };

    const state = {
      options: []
    };

    expect(QuestionReducer(state, action)).toMatchSnapshot();
  });

  test('DELETE_QO_SUCCESS', () => {
    const action = {
      type: QuestionConstants.DELETE_QO_SUCCESS,
      id: 1,
    };

    const state = {
      options: []
    };

    expect(QuestionReducer(state, action)).toMatchSnapshot();
  });

  test('EDIT_QUESTION_SUCCESS', () => {
    const action = {
      type: QuestionConstants.EDIT_QUESTION_SUCCESS,
    };

    expect(QuestionReducer(undefined, action)).toMatchSnapshot();
  });
});