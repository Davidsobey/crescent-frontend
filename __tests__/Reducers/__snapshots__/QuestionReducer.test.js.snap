// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionReducer CREATE_SUCCESS 1`] = `
Object {
  "creating": false,
  "question": Object {},
}
`;

exports[`QuestionReducer DELETE_QO_SUCCESS 1`] = `
Object {
  "options": Array [],
  "options_loading": false,
}
`;

exports[`QuestionReducer DELETE_SUCCESS 1`] = `
Object {
  "loading": false,
  "questions": Array [],
}
`;

exports[`QuestionReducer EDIT_QUESTION_SUCCESS 1`] = `
Object {
  "question": undefined,
  "question_editing": false,
}
`;

exports[`QuestionReducer GETALL_SUCCESS 1`] = `
Object {
  "loading": false,
  "questions": Array [],
}
`;

exports[`QuestionReducer GETBYID_SUCCESS 1`] = `
Object {
  "question": Object {
    "id": 1,
    "name": "111",
  },
  "question_loading": false,
}
`;

exports[`QuestionReducer LOADQUESTIONS_SUCCESS 1`] = `
Object {
  "loading": false,
  "questions": Array [],
}
`;

exports[`QuestionReducer OPTIONS_ISANSWER_SUCCESS 1`] = `
Object {
  "options": Object {
    "0": Object {
      "isAnswer": true,
    },
  },
  "options_loading": false,
}
`;

exports[`QuestionReducer OPTIONS_SUCCESS 1`] = `
Object {
  "options": Array [],
  "options_loading": false,
}
`;
